File: scala/Sorting_algorithms/quicksort.scala
object Quicksort {
    def quicksort[T](cmp: (T,T) => Boolean, list: List[T]) : List[T] = {
        list match {
            case Nil =>
                Nil
            case pivot::tl =>
                val (e1, e2) = tl.partition( (X: T) => cmp(X, pivot) )
                quicksort(cmp, e1) ::: (pivot::Nil) ::: quicksort(cmp, e2)
        }
    }
}

File: edition/2017/erlang/Sorting_algorithm/quicksort.erl
-module(quicksort).
-export([quicksort/2]).

quicksort(_, []) ->
    [];
quicksort(P, [Pivot | T]) ->
    quicksort(P, [X || X <- T, P(X, Pivot)]) ++ [Pivot] ++ quicksort(P, [X || X <- T, not P(X, Pivot)]).
